code/datums/components/shell.dm

```diff
@@ -0,0 +1,178 @@
+/// Makes an atom a shell that is able to take in an attached circuit.
+/datum/component/shell
+	dupe_mode = COMPONENT_DUPE_UNIQUE
+
+	/// The circuitboard attached to this shell
+	var/obj/item/integrated_circuit/attached_circuit
+
+	/// Flags containing what this shell can do
+	var/shell_flags = 0
+
+	/// The capacity of the shell.
+	var/capacity = INFINITY
+
+	/// A list of components that cannot be removed
+	var/list/obj/item/circuit_component/unremovable_components
+
+/datum/component/shell/Initialize(unremovable_components, capacity = src.capacity, shell_flags = src.shell_flags)
```

Comment:
Don't use default parameters like this, I believe we've had problems with BYOND in the past about it.

Instead, just do `src.capacity = capacity || src.capacity`, or an if-statement.